File: src/main/java/com/example/Sapo_Clone/Entity/Promotion.java
package com.example.Sapo_Clone.Entity;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import lombok.experimental.FieldDefaults;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "promotions")
@Data
@AllArgsConstructor
@NoArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Promotion {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    int id;

    @Column(name = "promotion_name", columnDefinition = "NVARCHAR(255)")
    String promotionName;

    @Column(name = "description", columnDefinition = "TEXT")
    String description;

    @Column(name = "scope")
    int scope; // 0: Invoice, 1: Specific Product

    @Column(name = "discount_type")
    int discountType; // 0: Flat amount, 1: Percentage

    @Column(name = "discount_value")
    double discountValue; // Ex: 10% or 50,000 VND

    @Column(name = "min_account")
    double minAccount; // Minimum purchase required to apply

    @Column(name = "max_account")
    double maxAccount; // Max discount allowed for percentage (e.g. 50k for 10%)

    @Column(name = "status")
    int status; // 0: Inactive, 1: Active, 2: Outdated

    @Column(name = "created_at")
    LocalDateTime createdAt;

    @Column(name = "started_at")
    LocalDateTime startedAt;

    @Column(name = "ended_at")
    LocalDateTime endedAt;

    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @ManyToOne
    @JoinColumn(name = "company_id")
    Company company;

    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @ManyToMany
    @JoinTable(name = "promotion_product", joinColumns = @JoinColumn(name = "promotion_id"), inverseJoinColumns = @JoinColumn(name = "product_id"))
    List<Product> products;

    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @OneToMany(mappedBy = "promotion")
    List<Order> orders;

    @PrePersist
    public void onCreate() {
        createdAt = LocalDateTime.now();
    }
}

File: lib/models/promotion.dart
class PromotionResponse {
  final int id;
  final int scope; // 0: Order, 1: Product
  final String promotionName;
  final double discountValue;
  final String discountType; // "0" for Flat, "1" for Percentage
  final double? maxAccount;
  final double? minAccount;
  final String? description;
  final String? startDate;
  final String? endDate;
  final int status;
  final List<int>? productIds;

  PromotionResponse({
    required this.id,
    required this.scope,
    required this.promotionName,
    required this.discountValue,
    required this.discountType,
    this.maxAccount,
    this.minAccount,
    this.description,
    this.startDate,
    this.endDate,
    required this.status,
    this.productIds,
  });

  DateTime? get startedAt => startDate != null ? DateTime.tryParse(startDate!) : null;
  DateTime? get endedAt => endDate != null ? DateTime.tryParse(endDate!) : null;

  factory PromotionResponse.fromJson(Map<String, dynamic> json) {
    return PromotionResponse(
      id: json['id'] as int,
      scope: json['scope'] as int,
      promotionName: json['promotionName'] as String,
      discountValue: (json['discountValue'] as num).toDouble(),
      discountType: json['discountType'] as String,
      maxAccount: (json['maxAccount'] as num?)?.toDouble(),
      minAccount: (json['minAccount'] as num?)?.toDouble(),
      description: json['description'] as String?,
      startDate: json['startDate'] as String?,
      endDate: json['endDate'] as String?,
      status: json['status'] as int,
      productIds: (json['productIds'] as List<dynamic>?)?.map((e) => e as int).toList(),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'scope': scope,
    'promotionName': promotionName,
    'discountValue': discountValue,
    'discountType': discountType,
    'maxAccount': maxAccount,
    'minAccount': minAccount,
    'description': description,
    'startDate': startDate,
    'endDate': endDate,
    'status': status,
    'productIds': productIds,
  };
}

class PromotionListResponse {
  final int id;
  final int scope;
  final String promotionName;
  final double discountValue;
  final String discountType; // "0" or "1"
  final double? maxAccount;
  final double? minAccount;
  final String? description;
  final String? startDate;
  final String? endDate;
  final int status;

  PromotionListResponse({
    required this.id,
    required this.scope,
    required this.promotionName,
    required this.discountValue,
    required this.discountType,
    this.maxAccount,
    this.minAccount,
    this.description,
    this.startDate,
    this.endDate,
    required this.status,
  });

  DateTime? get startedAt => startDate != null ? DateTime.tryParse(startDate!) : null;
  DateTime? get endedAt => endDate != null ? DateTime.tryParse(endDate!) : null;

  factory PromotionListResponse.fromJson(Map<String, dynamic> json) {
    return PromotionListResponse(
      id: json['id'] as int,
      scope: json['scope'] as int,
      promotionName: json['promotionName'] as String,
      discountValue: (json['discountValue'] as num).toDouble(),
      discountType: json['discountType'] as String,
      maxAccount: (json['maxAccount'] as num?)?.toDouble(),
      minAccount: (json['minAccount'] as num?)?.toDouble(),
      description: json['description'] as String?,
      startDate: json['startDate'] as String?,
      endDate: json['endDate'] as String?,
      status: json['status'] as int,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'scope': scope,
    'promotionName': promotionName,
    'discountValue': discountValue,
    'discountType': discountType,
    'maxAccount': maxAccount,
    'minAccount': minAccount,
    'description': description,
    'startDate': startDate,
    'endDate': endDate,
    'status': status,
  };
}
